File: solutions/01-introductory-problems/apple-division.py
# https://cses.fi/problemset/task/1623

def sol(c: list[int], diff: int=0) -> int:
    if not c:
        return diff

    return min([abs(sol(c[1:], diff-c[0])),
                abs(sol(c[1:], diff+c[0]))])


n = int(input())
arr=list(map(int, input().split(" ")))

print(sol(arr))

File: solutions/01-introductory-problems/apple-division.cpp
// https://cses.fi/problemset/task/1623
#include <bits/stdc++.h>
typedef long long ll;

ll solve(std::vector<int> &arr, size_t i, ll diff)
{
    if (arr.size()==i) return diff;
    i++;
    return std::min(
        std::abs(solve(arr,i,diff+arr[i-1])),
        std::abs(solve(arr,i,diff-arr[i-1])));
}


int main()
{
    int n; std::cin>>n;
    std::vector<int> arr(n);
    for(int i =0; i<n;i++) std::cin >> arr[i];
    std::cout << solve(arr,0, 0) << std::endl;
    return 0;
}